File: Museo/Clases/AgenteRestaurador.py
from __future__ import annotations

import os
import sys

sys.path.insert(0, os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))

from datetime import date

from Clases.Obra import EstadoObra
from Clases.Usuario import Usuario


class AgenteRestaurador(Usuario):
    """Usuario encargado de gestionar las restauraciones de las obras."""

    # --------------------Metodos--------------------------------------------------

    def get_rol(self) -> str:
        """Retorna el rol del usuario."""
        return "Agente Restaurador"

    def iniciar_restauracion(self, obra) -> None:
        """Inicia una restauración sobre una obra y cambia su estado a EN_RESTAURACION."""
        if not self._autenticado:
            raise PermissionError("Debe autenticarse para iniciar restauraciones.")
        restauracion = {
            "fecha_inicio": date.today(),
            "fecha_fin": None,
        }
        obra.restauraciones.append(restauracion)
        obra._estado = EstadoObra.EN_RESTAURACION

    def finalizar_restauracion(self, obra) -> None:
        """Finaliza la restauración activa de una obra y registra la fecha de fin."""
        if not self._autenticado:
            raise PermissionError("Debe autenticarse para finalizar restauraciones.")
        restauracion_activa = next(
            (r for r in obra.restauraciones if r["fecha_fin"] is None), None
        )
        if restauracion_activa is None:
            raise ValueError(f"La obra '{obra.titulo}' no tiene una restauración activa.")
        restauracion_activa["fecha_fin"] = date.today()
        obra._estado = EstadoObra.EXPUESTA

    def consultar_restauraciones(self, obra) -> list:
        """Retorna el historial de restauraciones de una obra, ordenado por antigüedad."""
        if not self._autenticado:
            raise PermissionError("Debe autenticarse para consultar restauraciones.")
        return sorted(obra.restauraciones, key=lambda r: r["fecha_inicio"])

    def verificar_obras_pendientes(self, obras: list) -> list:
        """Retorna las obras que necesitan restauración según el criterio de 5 años."""
        if not self._autenticado:
            raise PermissionError("Debe autenticarse para verificar obras pendientes.")
        return [obra for obra in obras if obra.necesita_restauracion()]


# ---------------------------------------------------------------------------
# Prueba interactiva
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import io
    from datetime import date

    from Clases.Cuadro import Cuadro, EstiloCuadro, TecnicaCuadro
    from Clases.Obra import Periodo

    # Silenciar creación de objetos
    _stdout = sys.stdout
    sys.stdout = io.StringIO()

    restaurador = AgenteRestaurador("restaurador01", "rest456")

    cuadro = Cuadro(
        "La Gioconda", "Leonardo da Vinci", Periodo.RENACIMIENTO,
        date(1503, 1, 1), date(1797, 8, 1),
        EstiloCuadro.RENACIMIENTO, TecnicaCuadro.OLEO,
    )

    sys.stdout = _stdout

    print("=== Prueba de AgenteRestaurador ===")
    print(f"Restaurador: {restaurador.nombre_usuario} | Rol: {restaurador.get_rol()}")

    restaurador.autenticar("rest456")
    print(f"Autenticado: {restaurador.autenticado}")
    print(f"\nObra: {cuadro.titulo} | Estado: {cuadro.estado.value}")
    print(f"¿Necesita restauración? {cuadro.necesita_restauracion()}")

    input("\nPresiona ENTER para iniciar restauración...")
    restaurador.iniciar_restauracion(cuadro)
    print(f"Estado: {cuadro.estado.value}")

    input("\nPresiona ENTER para finalizar restauración...")
    restaurador.finalizar_restauracion(cuadro)
    print(f"Estado: {cuadro.estado.value}")

    historial = restaurador.consultar_restauraciones(cuadro)
    print(f"\nHistorial de restauraciones: {historial}")

File: 2.2Museo_Java/Clases_Museo/AgenteRestaurador.java
package Clases_Museo;

import java.time.LocalDate;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

/**
 * Usuario encargado de gestionar las restauraciones de las obras.
 *
 * Extiende Usuario con operaciones para iniciar y finalizar
 * restauraciones, consultar su historial y detectar obras pendientes.
 */
public class AgenteRestaurador extends Usuario {

    // --- Constructor ---

    /**
     * Inicializa el agente restaurador con sus credenciales de acceso.
     *
     * @param nombreUsuario Nombre de usuario del sistema.
     * @param contrasena    Contrasena del agente.
     */
    public AgenteRestaurador(String nombreUsuario, String contrasena) {
        super(nombreUsuario, contrasena);
    }

    // --- Metodos ---

    /** @return El rol del usuario en el sistema. */
    @Override
    public String getRol() { return "Agente Restaurador"; }

    /**
     * Inicia una restauracion sobre una obra y cambia su estado a EN_RESTAURACION.
     *
     * @param obra Obra sobre la que se iniciara la restauracion.
     * @throws SecurityException si el agente no esta autenticado.
     */
    public void iniciarRestauracion(Obra obra) {
        if (!autenticado) {
            throw new SecurityException("Debe autenticarse para iniciar restauraciones.");
        }
        Map<String, Object> restauracion = new HashMap<>();
        restauracion.put("fechaInicio", LocalDate.now());
        restauracion.put("fechaFin", null);
        obra.restauraciones.add(restauracion);
        obra.estado = Obra.EstadoObra.EN_RESTAURACION;
    }

    /**
     * Finaliza la restauracion activa de una obra y registra la fecha de fin.
     *
     * @param obra Obra con restauracion en curso.
     * @throws SecurityException        si el agente no esta autenticado.
     * @throws IllegalArgumentException si la obra no tiene restauracion activa.
     */
    @SuppressWarnings("unchecked")
    public void finalizarRestauracion(Obra obra) {
        if (!autenticado) {
            throw new SecurityException("Debe autenticarse para finalizar restauraciones.");
        }
        Map<String, Object> activa = null;
        for (Object r : obra.restauraciones) {
            Map<String, Object> map = (Map<String, Object>) r;
            if (map.get("fechaFin") == null) {
                activa = map;
                break;
            }
        }
        if (activa == null) {
            throw new IllegalArgumentException(
                "La obra '" + obra.getTitulo() + "' no tiene una restauracion activa.");
        }
        activa.put("fechaFin", LocalDate.now());
        obra.estado = Obra.EstadoObra.EXPUESTA;
    }

    /**
     * Retorna el historial de restauraciones de una obra, ordenado por fecha de inicio.
     *
     * @param obra Obra de la que se consultara el historial.
     * @return Lista de restauraciones ordenada cronologicamente.
     * @throws SecurityException si el agente no esta autenticado.
     */
    @SuppressWarnings("unchecked")
    public List<Object> consultarRestauraciones(Obra obra) {
        if (!autenticado) {
            throw new SecurityException("Debe autenticarse para consultar restauraciones.");
        }
        return obra.restauraciones.stream()
            .sorted(Comparator.comparing(r ->
                (LocalDate) ((Map<String, Object>) r).get("fechaInicio")))
            .collect(Collectors.toList());
    }

    /**
     * Retorna las obras que nunca han sido restauradas.
     *
     * @param obras Lista de obras a verificar.
     * @return Lista de obras que necesitan restauracion.
     * @throws SecurityException si el agente no esta autenticado.
     */
    public List<Obra> verificarObrasPendientes(List<Obra> obras) {
        if (!autenticado) {
            throw new SecurityException("Debe autenticarse para verificar obras pendientes.");
        }
        return obras.stream()
            .filter(Obra::necesitaRestauracion)
            .collect(Collectors.toList());
    }
}
